extensions/amp-form/0.1/form-proxy.js

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+/**
+ * Creates a proxy object `form.$p` that proxies all of the methods and
+ * properties to the original DOM APIs. This is to work around the problematic
+ * forms feature where inputs mask DOM APIs.
+ *
+ * E.g. a `<input id="id">` will override `form.id` from the original DOM API.
+ * Form proxy will give access to the original `id` value via `form.$p.id`.
+ *
+ * @param {!HTMLFormElement} form
+ * @return {!Object}
+ */
+export function installFormProxy(form) {
+  const constr = getFormProxyConstr(form.ownerDocument.defaultView);
+  const proxy = new constr(form);
+  form['$p'] = proxy;
+  return proxy;
+}
+
+
+/**
+ * @param {!Window} win
+ * @return {function(new:!Object)}
+ */
+function getFormProxyConstr(win) {
+  if (!win.FormProxy) {
+    win.FormProxy = createFormProxyConstr(win);
+  }
+  return win.FormProxy;
+}
+
+
+/**
+ * @param {!Window} win
+ * @return {function(new:!Object)}
+ */
+function createFormProxyConstr(win) {
+
+  /**
+   * @param {!HTMLFormElement} form
+   * @constructor
+   */
+  function FormProxy(form) {
+    /** @private @const {!HTMLFormElement} */
+    this.form = form;
+  }
+
+  const FormProxyProto = FormProxy.prototype;
+
+  // Hierarchy:
+  //   Node  <==  Element <== HTMLElement <== HTMLFormElement
+  //   EventTarget  <==  HTMLFormElement
+  const prototypes = [
+    win.HTMLFormElement.prototype,
+    win.HTMLElement.prototype,
+    win.Element.prototype,
+    win.Node.prototype,
+    win.EventTarget.prototype,
+  ];
+  prototypes.forEach(function(prototype) {
+    const properties = win.Object.getOwnPropertyDescriptors(prototype);
+    for (const name in properties) {
+      if (win.Object.hasOwnProperty(FormProxyProto, name) ||
+          name == 'constructor') {
```

Comment:
This is redundant, it'll never be true without the proceeding conditional being true.